rasa/utils/tensorflow/temp_keras_modules.py

```diff
@@ -2,28 +2,50 @@
 from typing import List, Dict, Union, Optional, Any, Generator, Tuple, Iterator
 
 import numpy as np
+
 import tensorflow as tf
 
-from keras.callbacks import Callback, History
-from keras.engine import training, training_utils, data_adapter, base_layer
-from keras.engine.data_adapter import DataHandler
+# Note: the below is the same as `from keras.engine.training import Model`
+from tensorflow.keras import Model
+
+# Note: the following import is from the temporary/non-public tf.python module but
+# should stay since it's used in the current reference implementation (see link below)
 from tensorflow.python.eager import context
-from keras.utils import tf_utils, version_utils
-from tensorflow.python.profiler import trace
+
+# Note: the following imports are  from keras directly since those are the imports
+# used by the reference implementation:
+# https://github.com/keras-team/keras/blob/v2.7.0/keras/engine/training.py#L30
+from keras.engine import base_layer, training_utils, data_adapter
+from keras import callbacks as callbacks_module
+from keras.callbacks import Callback, History
+from keras.utils import traceback_utils, tf_utils, version_utils
+
+
+def _disallow_inside_tf_function(method_name: str) -> None:
+    if tf.inside_function():
+        error_msg = (
+            "Detected a call to `Model.{method_name}` inside a `tf.function`. "
+            "`Model.{method_name} is a high-level endpoint that manages its own "
+            "`tf.function`. Please move the call to `Model.{method_name}` outside "
+            "of all enclosing `tf.function`s. Note that you can call a `Model` "
+            "directly on `Tensor`s inside a `tf.function` like: `model(x)`."
+        ).format(method_name=method_name)
+        raise RuntimeError(error_msg)
 
 
 # noinspection PyMethodOverriding
-class TmpKerasModel(tf.keras.models.Model):
+class TmpKerasModel(Model):
     """Temporary solution. Keras model that uses a custom data adapter inside fit."""
 
     # TODO
-    #  we don't need this anymore once
+    #  we don't need this anymore once the fix from
     #  https://github.com/tensorflow/tensorflow/pull/45338
-    #  is merged and released
+    #  has been ported over to keras and merged there
 
     # This code is adapted from
-    # https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/python/keras/engine/training.py#L824-L1146
+    # https://github.com/keras-team/keras/blob/v2.7.0/keras/engine/training.py#L65
```

Comment:
Since its really just the `fit` function, this line directly takes you to the function 😬 

```suggestion
    # https://github.com/keras-team/keras/blob/v2.7.0/keras/engine/training.py#L902
```